Key store manager+FileSystem.h simulator

deuxksix opened the pull request:
This PR does two things:

1. add support of `src/system/physical/fileSystem.h` in simulator
2. add abstraction to give everyone access to the common key-value store

-------------

Users may discover new static methods in their mode context:

![2024-12-02_17-31](https://github.com/user-attachments/assets/5b57b92e-dc69-480e-8d7a-fa0b9dbeb28d)

-------------

There are pointers to the `KeyProxy` class which describes a typical use case:

![2024-12-02_17-32](https://github.com/user-attachments/assets/dcff863b-4f88-4e24-b95c-8e68b067e60d)

-------------

Others may discover these functionnalities through the basic mode documentation:

![2024-12-02_17-33](https://github.com/user-attachments/assets/21f2b287-918b-4dc4-8e9d-470447442029)

**Note: intentionally, we try to orient users into using the custom ramp & index, more basic & less error-prone!**

-------------

Other ways to discover direct access to the key-value store system is through its namespace:

![2024-12-02_17-34](https://github.com/user-attachments/assets/c0b39814-9721-46d7-a6d9-83c0482ef977)
![2024-12-02_17-35](https://github.com/user-attachments/assets/2d07370a-9d6d-466d-a51b-44366daf1e41)

-------------

Several remarks:

1. accessing key / values by "name" hashes the key name on the 31 lower-significant bits, prone to collisions
2. all other keys starts with a one, with the next 15 bits partitioning the key space in way to avoid collisions
3. the manager, each group and each mode have all their own `uint16_t` key space, indexed with enumerations
4. each space has a `storeId` serial, used to detect when groups / modes changes, dropping "obsolete" keys
5. values up to 64 bytes in size are transparently stored on several keys, only when using indexed key spaces

And as a bonus, at last, the lamp now remembers what was its last mode, its ramp value, other favorites, etc

**Draft: I still need to validate that everything works on real hardware!**

deuxksix commented:
draft status removed

now tested with a couple issues solved

BaptisteHudyma commented on src/modes/include/hardware/keystore.hpp:
```diff
@@ -0,0 +1,344 @@
+    else if constexpr (sizeof(valueTy) <= sizeof(uint32_t))
+    {
+      uint32_t storage = 0;
+      memcpy(&storage, &value, sizeof(value));
```
I'm 90% sure that in cpp we have better options than memcpy, but none comes to mind right now

deuxksix replied:
In a pure C++ fashion, doing `std::reinterpret_cast` is not the right way of doing object representation change, contrary to a common belief among C++ developers (I wonder where such misinformation comes from)

The modern way of doing that is [`std::bit_cast`](https://en.cppreference.com/w/cpp/numeric/bit_cast) from C++20 which is explicitly defined as:
```c++
template<class To, class From>
std::enable_if_t<
    sizeof(To) == sizeof(From) &&
    std::is_trivially_copyable_v<From> &&
    std::is_trivially_copyable_v<To>,
    To>
// constexpr support needs compiler magic
bit_cast(const From& src) noexcept
{
    static_assert(std::is_trivially_constructible_v<To>,
        "This implementation additionally requires "
        "destination type to be trivially constructible");
 
    To dst;
    std::memcpy(&dst, &src, sizeof(To));
    return dst;
}
```

We are C++17 here, and yes I've yolo-ed the `std::is_trivially_copyable` and `std::is_trivially_constructible` criteria in the aforementioned code, but 90% of the time it should be safe for the end-user to use as-is.

What I can do, to reassure the audience, is to integrate the `std::bit_cast` implementation quoted above in our `modes/include/tools.hpp` toolbox, with a link to the [cppreference associated page](https://en.cppreference.com/w/cpp/numeric/bit_cast) to be more explicit.

BaptisteHudyma commented:
Looks awesome thanks!!!
I will test it on several hardware before merge

deuxksix commented on src/modes/include/hardware/keystore.hpp:
```diff
@@ -0,0 +1,344 @@
+  assert((offset >> 4) == 0);    // offset on 4 bits (0-15)
+  assert((index >> 16) == 0);    // index on 16 bits (0-65535)
+
+  index |= indexKeyMask;
```
I can agree on
```diff
- index |= indexKeyMask;
+ index |= indexKeyFlag;
```

will propagate the change

deuxksix replied:
```diff
diff --git a/src/modes/include/hardware/keystore.hpp b/src/modes/include/hardware/keystore.hpp
index ed6f2c6..3abc219 100644
--- a/src/modes/include/hardware/keystore.hpp
+++ b/src/modes/include/hardware/keystore.hpp
@@ -102,15 +102,15 @@ static inline void LMBD_INLINE migrateIfNeeded()
   }
 }
 
-/** \private Mask common to all indexed keys
+/** \private Flag indexed / enumeration-based key space
  *
- * There are two ways of storing and retrieving persistent values:
+ * There are two ways of storing and retrieving persistent key store values:
  *  - by named key, a C-string hashed on the 31 least-significant bytes
  *  - by indexed keys, using fixed values, all starting by a 1 on their MSB
  *
- * This mask (0x8000000) reflects this and is or-ed to each PrefixValues
+ * This flag (0x8000000) reflects this and is or-ed to each PrefixValues
  */
-static constexpr uint32_t indexKeyMask = (1 << 31);
+static constexpr uint32_t indexKeyFlag = (1 << 31);
 
 /** \private Possible prefixes to differentiate indexed keys use cases
  *
@@ -130,10 +130,10 @@ static constexpr uint32_t indexKeyMask = (1 << 31);
  */
 enum class PrefixValues : uint32_t
 {
-  generalKeys = indexKeyMask | (0x0 << 29), ///< General keys (user)
-  managerKeys = indexKeyMask | (0x1 << 29), ///< Manager keys (per-mode)
-  privateKeys = indexKeyMask | (0x2 << 29), ///< Private keys (system-only)
-  invalidKeys = indexKeyMask | (0x3 << 29), ///< (reserved) for later use
+  generalKeys = indexKeyFlag | (0x0 << 29), ///< General keys (user)
+  managerKeys = indexKeyFlag | (0x1 << 29), ///< Manager keys (per-mode)
+  privateKeys = indexKeyFlag | (0x2 << 29), ///< Private keys (system-only)
+  invalidKeys = indexKeyFlag | (0x3 << 29), ///< (reserved) for later use
 };
 
 /// \private Used for no group index (global to all groups and all modes)
@@ -154,7 +154,7 @@ indexKeyFor(PrefixValues prefix, uint32_t groupId, uint32_t modeId, uint32_t off
   assert((offset >> 4) == 0);    // offset on 4 bits (0-15)
   assert((index >> 16) == 0);    // index on 16 bits (0-65535)
 
-  index |= indexKeyMask;
+  index |= indexKeyFlag;
   return (rawPrefix << 29) | (groupId << 25) | (modeId << 20) | (offset << 16) | index;
 }
 ```

deuxksix commented on src/modes/include/hardware/keystore.hpp:
```diff
@@ -0,0 +1,344 @@
+{
+  uint32_t rawPrefix = (uint32_t)prefix;
+
+  assert((rawPrefix << 3) == 0); // prefix on 2 bits
```
```diff
diff --git a/src/modes/include/hardware/keystore.hpp b/src/modes/include/hardware/keystore.hpp
index 376208f..c681da6 100644
--- a/src/modes/include/hardware/keystore.hpp
+++ b/src/modes/include/hardware/keystore.hpp
@@ -148,14 +148,13 @@ indexKeyFor(PrefixValues prefix, uint32_t groupId, uint32_t modeId, uint32_t off
 {
   uint32_t rawPrefix = (uint32_t)prefix;
 
-  assert((rawPrefix << 3) == 0); // prefix on 2 bits
-  assert((groupId >> 4) == 0);   // groupId on 4 bits (0-14)
-  assert((modeId >> 5) == 0);    // modeId on 5 bits (0-30)
-  assert((offset >> 4) == 0);    // offset on 4 bits (0-15)
-  assert((index >> 16) == 0);    // index on 16 bits (0-65535)
-
-  index |= indexKeyMask;
-  return (rawPrefix << 29) | (groupId << 25) | (modeId << 20) | (offset << 16) | index;
+  assert((rawPrefix >> 31) == 0b1); // prefix on 2 of the MSBs +indexKeyFlag
+  assert((groupId >> 4) == 0);      // groupId on 4 bits (0-14)
+  assert((modeId >> 5) == 0);       // modeId on 5 bits (0-30)
+  assert((offset >> 4) == 0);       // offset on 4 bits (0-15)
+  assert((index >> 16) == 0);       // index on 16 bits (0-65535)
+
+  return rawPrefix | (groupId << 25) | (modeId << 20) | (offset << 16) | index;
 }
 
 /// \private Orchestrate interactions with the indexed key store
```

fixed the code, it was bogus sorry!

deuxksix commented:
rebased +also fixed a bug

**Note:** it could be cool at some point to have explicit `enter_mode` / `quit_mode` / `enter_group` / `quit_group` handlers because right now, as there are several ways to enter / quit things (turn on, turn off, next group, next mode, favorite, etc) it's easy to forget to load / save something when writing bugs (see `save_ramps` called in three distinct places)

BaptisteHudyma commented:
Tested and approved, congrats <3
